File: java/src/main/java/vsdk/toolkit/galerkin/Interaction.java
package vsdk.toolkit.galerkin;

import vsdk.toolkit.common.Error;

public class Interaction {
    private static int totalInteractions;
    private static int ccInteractions;
    private static int csInteractions;
    private static int scInteractions;
    private static int ssInteractions;

    public GalerkinElement receiverElement;
    public GalerkinElement sourceElement;
    public float[] K; // Coupling coefficient(s), stored top to bottom, left to right
    public float[] deltaK; // Used for approximation error estimation over the link
    public byte numberOfBasisFunctionsOnReceiver;
    public byte numberOfBasisFunctionsOnSource;
    public byte numberOfReceiverCubaturePositions;
    public byte visibility; // 255 for full visibility, 0 for full occlusion

    public Interaction() {
    }

    public Interaction(
        GalerkinElement inReceiverElement,
        GalerkinElement inSourceElement,
        float[] inK,
        float[] inDeltaK,
        byte inNumberOfBasisFunctionsOnReceiver,
        byte inNumberOfBasisFunctionsOnSource,
        byte inNumberOfReceiverCubaturePositions,
        byte inVisibility
    ) {
        receiverElement = inReceiverElement;
        sourceElement = inSourceElement;
        numberOfBasisFunctionsOnReceiver = inNumberOfBasisFunctionsOnReceiver;
        numberOfBasisFunctionsOnSource = inNumberOfBasisFunctionsOnSource;
        numberOfReceiverCubaturePositions = inNumberOfReceiverCubaturePositions;
        visibility = inVisibility;

        int n = (numberOfBasisFunctionsOnReceiver & 0xFF) * (numberOfBasisFunctionsOnSource & 0xFF);
        if ( n <= 0 ) {
            n = 1;
        }
        K = new float[n];
        if ( inK != null ) {
            System.arraycopy(inK, 0, K, 0, Math.min(inK.length, K.length));
        }

        if ( (numberOfReceiverCubaturePositions & 0xFF) > 1 ) {
            Error.fatal(2, "interactionCreate", "Not yet implemented for higher order approximations");
        }

        deltaK = new float[1];
        if ( inDeltaK != null && inDeltaK.length > 0 ) {
            deltaK[0] = inDeltaK[0];
        }

        totalInteractions++;
        if ( inReceiverElement != null && inReceiverElement.isCluster() ) {
            if ( inSourceElement != null && inSourceElement.isCluster() ) {
                ccInteractions++;
            }
            else {
                scInteractions++;
            }
        }
        else {
            if ( inSourceElement != null && inSourceElement.isCluster() ) {
                csInteractions++;
            }
            else {
                ssInteractions++;
            }
        }
    }

    public static int getNumberOfInteractions() {
        return totalInteractions;
    }

    public static int getNumberOfClusterToClusterInteractions() {
        return ccInteractions;
    }

    public static int getNumberOfClusterToSurfaceInteractions() {
        return csInteractions;
    }

    public static int getNumberOfSurfaceToClusterInteractions() {
        return scInteractions;
    }

    public static int getNumberOfSurfaceToSurfaceInteractions() {
        return ssInteractions;
    }

    public static void interactionDestroy(Interaction interaction) {
        if ( interaction == null ) {
            return;
        }

        totalInteractions--;
        if ( interaction.receiverElement != null && interaction.receiverElement.isCluster() ) {
            if ( interaction.sourceElement != null && interaction.sourceElement.isCluster() ) {
                ccInteractions--;
            }
            else {
                scInteractions--;
            }
        }
        else {
            if ( interaction.sourceElement != null && interaction.sourceElement.isCluster() ) {
                csInteractions--;
            }
            else {
                ssInteractions--;
            }
        }
    }

    public static Interaction interactionDuplicate(Interaction interaction) {
        if ( interaction == null ) {
            return null;
        }
        return new Interaction(
            interaction.receiverElement,
            interaction.sourceElement,
            interaction.K,
            interaction.deltaK,
            interaction.numberOfBasisFunctionsOnReceiver,
            interaction.numberOfBasisFunctionsOnSource,
            interaction.numberOfReceiverCubaturePositions,
            interaction.visibility
        );
    }
}

File: typescript/src/main/vsdk/toolkit/galerkin/Interaction.ts
import { Error as VsdkError } from "../common/Error";
import { GalerkinElement } from "./GalerkinElement";

export class Interaction {
  private static totalInteractions = 0;
  private static ccInteractions = 0;
  private static csInteractions = 0;
  private static scInteractions = 0;
  private static ssInteractions = 0;

  public receiverElement!: GalerkinElement;
  public sourceElement!: GalerkinElement;
  public K!: number[];
  public deltaK!: number[];
  public numberOfBasisFunctionsOnReceiver!: number;
  public numberOfBasisFunctionsOnSource!: number;
  public numberOfReceiverCubaturePositions!: number;
  public visibility!: number;

  public constructor();
  public constructor(
    inReceiverElement: GalerkinElement,
    inSourceElement: GalerkinElement,
    inK: number[] | null,
    inDeltaK: number[] | null,
    inNumberOfBasisFunctionsOnReceiver: number,
    inNumberOfBasisFunctionsOnSource: number,
    inNumberOfReceiverCubaturePositions: number,
    inVisibility: number,
  );
  public constructor(
    inReceiverElement?: GalerkinElement,
    inSourceElement?: GalerkinElement,
    inK?: number[] | null,
    inDeltaK?: number[] | null,
    inNumberOfBasisFunctionsOnReceiver?: number,
    inNumberOfBasisFunctionsOnSource?: number,
    inNumberOfReceiverCubaturePositions?: number,
    inVisibility?: number,
  ) {
    this.K = [];
    this.deltaK = [0.0];
    this.numberOfBasisFunctionsOnReceiver = 0;
    this.numberOfBasisFunctionsOnSource = 0;
    this.numberOfReceiverCubaturePositions = 0;
    this.visibility = 0;

    if (arguments.length === 0) {
      return;
    }

    this.receiverElement = (inReceiverElement ?? null) as unknown as GalerkinElement;
    this.sourceElement = (inSourceElement ?? null) as unknown as GalerkinElement;
    this.numberOfBasisFunctionsOnReceiver = inNumberOfBasisFunctionsOnReceiver ?? 0;
    this.numberOfBasisFunctionsOnSource = inNumberOfBasisFunctionsOnSource ?? 0;
    this.numberOfReceiverCubaturePositions = inNumberOfReceiverCubaturePositions ?? 0;
    this.visibility = inVisibility ?? 0;

    let n = (this.numberOfBasisFunctionsOnReceiver & 0xff) * (this.numberOfBasisFunctionsOnSource & 0xff);
    if (n <= 0) {
      n = 1;
    }
    this.K = new Array<number>(n).fill(0.0);
    if (inK != null) {
      for (let i = 0; i < Math.min(inK.length, this.K.length); i++) {
        this.K[i] = inK[i];
      }
    }

    if ((this.numberOfReceiverCubaturePositions & 0xff) > 1) {
      VsdkError.fatal(2, "interactionCreate", "Not yet implemented for higher order approximations");
    }

    this.deltaK = new Array<number>(1).fill(0.0);
    if (inDeltaK != null && inDeltaK.length > 0) {
      this.deltaK[0] = inDeltaK[0];
    }

    Interaction.totalInteractions++;
    if (this.receiverElement !== null && this.receiverElement.isCluster()) {
      if (this.sourceElement !== null && this.sourceElement.isCluster()) {
        Interaction.ccInteractions++;
      }
      else {
        Interaction.scInteractions++;
      }
    }
    else {
      if (this.sourceElement !== null && this.sourceElement.isCluster()) {
        Interaction.csInteractions++;
      }
      else {
        Interaction.ssInteractions++;
      }
    }
  }

  public static getNumberOfInteractions(): number {
    return Interaction.totalInteractions;
  }

  public static getNumberOfClusterToClusterInteractions(): number {
    return Interaction.ccInteractions;
  }

  public static getNumberOfClusterToSurfaceInteractions(): number {
    return Interaction.csInteractions;
  }

  public static getNumberOfSurfaceToClusterInteractions(): number {
    return Interaction.scInteractions;
  }

  public static getNumberOfSurfaceToSurfaceInteractions(): number {
    return Interaction.ssInteractions;
  }

  public static interactionDestroy(interaction: Interaction | null): void {
    if (interaction === null) {
      return;
    }

    Interaction.totalInteractions--;
    if (interaction.receiverElement !== null && interaction.receiverElement.isCluster()) {
      if (interaction.sourceElement !== null && interaction.sourceElement.isCluster()) {
        Interaction.ccInteractions--;
      }
      else {
        Interaction.scInteractions--;
      }
    }
    else {
      if (interaction.sourceElement !== null && interaction.sourceElement.isCluster()) {
        Interaction.csInteractions--;
      }
      else {
        Interaction.ssInteractions--;
      }
    }
  }

  public static interactionDuplicate(interaction: Interaction | null): Interaction | null {
    if (interaction === null) {
      return null;
    }
    return new Interaction(
      interaction.receiverElement,
      interaction.sourceElement,
      interaction.K,
      interaction.deltaK,
      interaction.numberOfBasisFunctionsOnReceiver,
      interaction.numberOfBasisFunctionsOnSource,
      interaction.numberOfReceiverCubaturePositions,
      interaction.visibility,
    );
  }
}
